elyra/validation/validate.py

```diff
@@ -0,0 +1,424 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import json
+import os
+import re
+
+import networkx as nx
+
+from elyra import __version__ as elyra_version
+from elyra.pipeline.parser import PipelineParser
+from packaging import version
+from traitlets.config import SingletonConfigurable
+from typing import Dict, Optional
+
+
+class PipelineValidationManager(SingletonConfigurable):
+
+    def validate(self, root_dir: str, pipeline: Dict) -> 'ValidationResponse':
+        """
+        Validates the pipeline JSON payload
+        :param root_dir: the absolute base path of the current elyra workspace
+        :param pipeline: the pipeline definition to be validated
+        :return: ValidationResponse containing any and all issues discovered during the validation
+        """
+        response = ValidationResponse()
+
+        try:
+            pipeline = json.loads(json.dumps(pipeline))
+        except ValueError:
+            response.add_message(severity='1',
+                                 message_type="invalidJSON",
+                                 message="Invalid JSON detected, unable to continue.")
+            return response
+
+        self.validate_pipeline_structure(pipeline=pipeline, response=response)
```

Comment:
These sub-methods should be prefixed with `_` since they're not intended to be called publically.

Reply:
updated